src/service.js

```diff
@@ -217,6 +221,47 @@ export function fromClass(win, id, constructor) {
   return getServiceInternal(win, win, id, undefined, constructor);
 }
 
+/**
+ * Registers a service given a class to be used as implementation.
+ * @param {!Window} win
+ * @param {string} id of the service.
+ * @param {function(!Window):!Object=} opt_factory
+  *@param {function(new:Object, !Window)=} opt_constructor
+ * @param {boolean=} opt_instantiate Whether to immediately create the service
+ */
+export function registerServiceBuilder(win,
+                                       id,
+                                       opt_factory,
+                                       opt_constructor,
+                                       opt_instantiate) {
+  win = getTopWindow(win);
+  registerServiceInternal(win, win, id, opt_factory, opt_constructor);
+  if (opt_instantiate) {
+    getServiceInternal(win, win, id);
+  }
+}
+
+/**
+ * Returns a service and registers it given a class to be used as
+ * implementation.
+ * @param {!Node|!./service/ampdoc-impl.AmpDoc} nodeOrDoc
+ * @param {string} id of the service.
+  *@param {function(!./service/ampdoc-impl.AmpDoc):Object=} opt_factory
+ * @param {function(new:Object, !./service/ampdoc-impl.AmpDoc)=} opt_constructor
+ * @param {boolean=} opt_instantiate Whether to immediately create the service
+ */
+export function registerServiceBuilderForDoc(nodeOrDoc,
+                                             id,
+                                             opt_factory,
+                                             opt_constructor,
+                                             opt_instantiate) {
+  const ampdoc = getAmpdoc(nodeOrDoc);
+  const holder = getAmpdocServiceHolder(ampdoc);
+  registerServiceInternal(holder, ampdoc, id, opt_factory, opt_constructor);
+  if (opt_instantiate) {
+    getServiceInternal(holder, ampdoc, id);
```

Comment:
I don't think so. If the caller of registerServiceBuilder chose to only register and not instantiate the service, then I'm not sure we should instantiate it here if there's an existing service promise. What are your thoughts?